packages/shared/components/molecules/AssetIcon.svelte

```diff
@@ -1,24 +1,47 @@
 <script lang="typescript">
+    import { getIconColorFromString } from '@core/account'
     import { NetworkProtocol } from '@core/network'
     import { IAsset } from '@core/wallet'
-    import { isBright } from '@lib/helpers'
+    import { updatePersistedAsset } from '@core/wallet/stores'
+    import { getInitials, isBright } from '@lib/helpers'
     import { Icon, VerificationBadge } from 'shared/components'
+    import { onMount } from 'svelte'
 
     export let asset: IAsset
     export let large = false
     export let showVerificationBadge = false
 
+    const MAX_DISPLAYED_INITIALS = 5
+
     let icon: string
+    let assetIconColor: string
+    let assetIconBackgroundColor: string
+    let assetInitials: string
+    let assetIconWrapperWidth: number
 
-    $: assetIconColor = isBright(asset?.metadata?.primaryColor) ? 'gray-800' : 'white'
-    $: switch (asset?.metadata?.name?.toLocaleLowerCase()) {
-        case NetworkProtocol.IOTA:
-        case NetworkProtocol.Shimmer:
+    onMount(() => {
+        assetIconBackgroundColor = asset?.metadata?.primaryColor
+        if (
+            asset?.metadata?.name?.toLocaleLowerCase() === NetworkProtocol.IOTA ||
+            asset?.metadata?.name?.toLocaleLowerCase() === NetworkProtocol.Shimmer
+        ) {
             icon = asset?.metadata?.name?.toLocaleLowerCase()
-            break
-        default:
-            icon = 'tokens'
-    }
+        } else {
+            // Generate our own icon
+            // Icon content: unit or initials
+            // Icon color: deterministic color based on asset name in color range 500-800
+            assetInitials =
+                asset?.metadata?.unit?.slice(0, MAX_DISPLAYED_INITIALS) ??
+                getInitials(asset?.metadata?.name, MAX_DISPLAYED_INITIALS)
+            // If the primaryColor is not yet set, find it and persist it
+            if (!asset?.metadata?.primaryColor) {
+                const colorFromString = getIconColorFromString(asset?.metadata?.name)
+                updatePersistedAsset({ id: asset?.id, metadata: { ...asset?.metadata, primaryColor: colorFromString } })
+                assetIconBackgroundColor = colorFromString
```

Comment:
Lets move this to packages/shared/lib/core/wallet/helpers/buildPersistedAssetFromIrc30Metadata.ts

Reply:
updated